src/Analyser/NodeScopeResolver.php

```diff
@@ -1836,25 +1836,10 @@ function (MutatingScope $scope) use ($expr, $nodeCallback, $context): Expression
 			) {
 				$arrayArg = $expr->getArgs()[0]->value;
 				$arrayArgType = $scope->getType($arrayArg);
-				$scope = $scope->invalidateExpression($arrayArg);
 
-				$functionName = $functionReflection->getName();
-				$arrayArgType = TypeTraverser::map($arrayArgType, static function (Type $type, callable $traverse) use ($functionName): Type {
-					if ($type instanceof UnionType || $type instanceof IntersectionType) {
-						return $traverse($type);
-					}
-					if ($type instanceof ConstantArrayType) {
-						return $functionName === 'array_pop' ? $type->removeLast() : $type->removeFirst();
-					}
-					if ($type->isIterableAtLeastOnce()->yes()) {
-						return $type->toArray();
-					}
-					return $type;
-				});
-
-				$scope = $scope->assignExpression(
+				$scope = $scope->invalidateExpression($arrayArg)->assignExpression(
 					$arrayArg,
-					$arrayArgType,
+					$functionReflection->getName() === 'array_pop' ? $arrayArgType->popArray() : $arrayArgType->shiftArray(),
 					$scope->getNativeType($arrayArg),
```

Comment:
I feel like we should get the native array type and call the same method on it too.